core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,286 @@
+import io.substrait.expression.FunctionLookup;
+import io.substrait.expression.ImmutableExpression;
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.function.SimpleExtension;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.proto.AggregateRel;
+import io.substrait.proto.FetchRel;
+import io.substrait.proto.FilterRel;
+import io.substrait.proto.JoinRel;
+import io.substrait.proto.ProjectRel;
+import io.substrait.proto.ReadRel;
+import io.substrait.proto.SortField;
+import io.substrait.proto.SortRel;
+import io.substrait.type.ImmutableNamedStruct;
+import io.substrait.type.NamedStruct;
+import io.substrait.type.Type;
+import io.substrait.type.proto.FromProto;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Converts from proto to pojo rel representation TODO: AdvancedExtension, CrossJoin, Set */
+public class ProtoRelConverter {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ProtoRelConverter.class);
+
+  private final FunctionLookup lookup;
+  private final SimpleExtension.ExtensionCollection extensions;
+  private final Type rootType;
+
+  public ProtoRelConverter(
+      FunctionLookup lookup, SimpleExtension.ExtensionCollection extensions, Type rootType) {
+    this.lookup = lookup;
+    this.extensions = extensions;
+    this.rootType = rootType;
+  }
+
+  public Rel from(io.substrait.proto.Rel rel) {
+    io.substrait.proto.Rel.RelTypeCase relType = rel.getRelTypeCase();
+    switch (relType) {
+      case READ -> {
+        ReadRel readRel = rel.getRead();
+        if (readRel.hasVirtualTable()) {
+          return newVirtualTable(readRel);
+        } else if (readRel.hasNamedTable()) {
+          return newNamedScan(readRel);
+        } else { // FIXME: EmptyScan?
+          return newEmptyScan(readRel);
+        }
+      }
+      case FILTER -> {
+        return newFilter(rel.getFilter());
+      }
+      case FETCH -> {
+        return newFetch(rel.getFetch());
+      }
+      case AGGREGATE -> {
+        return newAggregate(rel.getAggregate());
+      }
+      case SORT -> {
+        return newSort(rel.getSort());
+      }
+      case JOIN -> {
+        return newJoin(rel.getJoin());
+      }
+      case PROJECT -> {
+        return newProject(rel.getProject());
+      }
+      default -> {
+        //                case SET:
+        //                case EXTENSION_SINGLE:
+        //                case EXTENSION_MULTI:
+        //                case EXTENSION_LEAF:
+        //                case CROSS:
+        //                case RELTYPE_NOT_SET:
+        throw new UnsupportedOperationException();
+      }
+    }
+  }
+
+  private Filter newFilter(FilterRel filterRel) {
+    Rel input = from(filterRel.getInput());
+    ProtoExpressionConverter expressionConverter =
+        new ProtoExpressionConverter(lookup, extensions, input.getRecordType());
+    Expression expression = expressionConverter.from(filterRel.getCondition());
```

Comment:
For now, `ProtoExpressionConverter` would not handle expression containing subquery.  

Later on, if we extend the support of subquery in `ProtoExpressionConverter`, we need pass in a `ProtoRelConverter` to `ProtoExpressionConverter` because of the nested structure of Rel / Expression.
 
Maybe add a `TODO` mark here.

Reply:
Added TODO